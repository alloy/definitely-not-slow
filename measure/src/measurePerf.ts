import * as os from 'os';
import * as fs from 'fs';
import * as path from 'path';
import { LanguageServiceHost, CompilerOptions, Node, SourceFile, LanguageService } from 'typescript';
import { performance, PerformanceObserver } from 'perf_hooks';
import { LanguageServiceMeasurement, PackageMeasurement } from './types';
import { stdDev, mean } from './utils';
const basePath = path.resolve(__dirname, '..');

export interface MeasurePerfOptions {
  packageName: string;
  definitelyTypedRootPath: string;
  iterations: number;
  ts: typeof import('typescript');
}

export function measurePerf({
  definitelyTypedRootPath,
  ts,
  iterations,
  packageName,
}: MeasurePerfOptions) {
  const packagePath = ensureExists(path.resolve(definitelyTypedRootPath, 'types', packageName));
  const testPaths = getTestFileNames(packagePath);
  const compilerOptions = getCompilerOptionsForPackage(packagePath);
  const host = createLanguageServiceHostForTestFiles(compilerOptions, testPaths);
  const languageService = ts.createLanguageService(host);
  const measurements = testPaths.reduce(({ completions, quickInfo }, testPath) => {
    const sourceFile = ts.createSourceFile(
      testPath,
      ts.sys.readFile(testPath)!,
      compilerOptions.target || ts.ScriptTarget.Latest);
    console.log(` - ${path.relative(definitelyTypedRootPath, testPath)}`);
    return {
      completions: completions.concat(measureAtEachIdentifier(sourceFile, iterations, 'completions', getCompletionsAtPosition)),
      quickInfo: quickInfo.concat(measureAtEachIdentifier(sourceFile, iterations, 'quick info', getQuickInfoAtPosition)),
    };
  }, {
    completions: [] as LanguageServiceMeasurement[],
    quickInfo: [] as LanguageServiceMeasurement[],
  });

  const program = ts.createProgram({ rootNames: testPaths, options: compilerOptions });
  program.getSemanticDiagnostics();

  const measurement: PackageMeasurement = {
    typeCount: (program as any).getTypeCount(),
    relationCacheSizes: (program as any).getRelationCacheSizes(),
    completions: measurements.completions,
    quickInfo: measurements.quickInfo,
  };

  return measurement;

  function createLanguageServiceHostForTestFiles(
    compilerOptions: CompilerOptions,
    testPaths: string[],
  ): LanguageServiceHost {
    let version = 0;
    return {
      directoryExists: ts.sys.directoryExists,
      getCompilationSettings: () => compilerOptions,
      getCurrentDirectory: ts.sys.getCurrentDirectory,
      getDefaultLibFileName: () => path.resolve(__dirname, '../node_modules/typescript/lib/lib.d.ts'),
      getNewLine: () => ts.sys.newLine,
      getScriptFileNames: () => testPaths,
      fileExists: ts.sys.fileExists,
      getDirectories: ts.sys.getDirectories,
      getScriptSnapshot: fileName => ts.ScriptSnapshot.fromString(ts.sys.readFile(ensureExists(fileName))!),
      getScriptVersion: () => (version++).toString(),
    };
  }

  function measureAtEachIdentifier(sourceFile: SourceFile, iterations: number, progressTitle: string, fn: (languageService: LanguageService, fileName: string, pos: number) => boolean | undefined) {
    const identifiers = getIdentifiers(sourceFile);

    // Warm-up
    fn(languageService, sourceFile.fileName, 0);

    const measurements: LanguageServiceMeasurement[] = [];
    identifiers.forEach((identifier, index) => {
      updateProgress(progressTitle, index, identifiers.length);
      const start = identifier.getStart(sourceFile);
      const lineAndCharacter = ts.getLineAndCharacterOfPosition(sourceFile, start);
      const durations: number[] = [];
      const measurement: LanguageServiceMeasurement = {
        fileName: path.relative(basePath, sourceFile.fileName),
        line: lineAndCharacter.line + 1,
        offset: lineAndCharacter.character + 1,
        identifierText: identifier.getText(sourceFile),
        start,
        end: identifier.getEnd(),
        duration: -1,
        iterations,
        standardDeviation: 1,
      };

      const observer = new PerformanceObserver((list, obs) => {
        const measurement = list.getEntriesByName('languageServiceMeasurement')[0];
        durations.push(measurement.duration);
        obs.disconnect();
      });

      for (let i = 0; i < iterations; i++) {
        observer.observe({ entryTypes: ['measure'] });
        fn(languageService, sourceFile.fileName, start);
      }

      measurement.duration = mean(durations);
      measurement.standardDeviation = stdDev(durations);
      measurements.push(measurement);
    });

    return measurements;
  }

  function getIdentifiers(sourceFile: SourceFile) {
    const identifiers: Node[] = [];
    ts.forEachChild(sourceFile, function visit(node) {
      if (ts.isIdentifier(node)) {
        identifiers.push(node);
      }
      else {
        ts.forEachChild(node, visit);
      }
    });
    return identifiers;
  }

  function getCompletionsAtPosition(languageService: LanguageService, fileName: string, pos: number): boolean {
    performance.mark('beforeCompletions');
    const completions = languageService.getCompletionsAtPosition(fileName, pos, undefined);
    performance.mark('afterCompletions');
    performance.measure('languageServiceMeasurement', 'beforeCompletions', 'afterCompletions');
    return !!completions && completions.entries.length > 0;
  }

  function getQuickInfoAtPosition(languageService: LanguageService, fileName: string, pos: number): boolean {
    performance.mark('beforeQuickInfo');
    const quickInfo = languageService.getQuickInfoAtPosition(fileName, pos);
    performance.mark('afterQuickInfo');
    performance.measure('languageServiceMeasurement', 'beforeQuickInfo', 'afterQuickInfo');
    return !!quickInfo;
  }

  function getTestFileNames(packagePath: string): string[] {
    const files = ts.sys.readDirectory(packagePath, [ts.Extension.Ts, ts.Extension.Tsx]);
    return files.filter(file => !file.endsWith(ts.Extension.Dts));
  }

  function getCompilerOptionsForPackage(packagePath: string): CompilerOptions {
    const tsConfigPath = ensureExists(path.resolve(packagePath, 'tsconfig.json'));
    const parsedCommandLine = ts.getParsedCommandLineOfConfigFile(tsConfigPath, {}, {
      fileExists: ts.sys.fileExists,
      getCurrentDirectory: ts.sys.getCurrentDirectory,
      readDirectory: ts.sys.readDirectory,
      readFile: ts.sys.readFile,
      useCaseSensitiveFileNames: ts.sys.useCaseSensitiveFileNames,
      onUnRecoverableConfigFileDiagnostic: diagnostic => {
        console.error(ts.formatDiagnostic(diagnostic, {
          getCanonicalFileName: fileName => fileName,
          getCurrentDirectory: ts.sys.getCurrentDirectory,
          getNewLine: () => ts.sys.newLine,
        }));
      },
    });

    return parsedCommandLine ? parsedCommandLine.options : ts.getDefaultCompilerOptions();
  }
}

function updateProgress(title: string, done: number, total: number) {
  const padDigits = total.toString().length - done.toString().length;
  if (done === total - 1) {
    if (process.stdout.clearLine && process.stdout.cursorTo) {
      process.stdout.clearLine();
      process.stdout.cursorTo(0);
      process.stdout.write(`   - ${title} (âœ”)`);
    }
    process.stdout.write(os.EOL);
  } else if (!done) {
    process.stdout.write(`   - ${title}`);
  } else if (process.stdout.clearLine && process.stdout.cursorTo) {
    process.stdout.clearLine();
    process.stdout.cursorTo(0);
    process.stdout.write(`   - ${title} ${' '.repeat(padDigits)}(${done}/${total} positions)`);
  }
}

function ensureExists(...pathNames: string[]): string {
  for (const pathName of pathNames) {
    if (fs.existsSync(pathName)) {
      return pathName;
    }
  }
  const pathNamesPrint = pathNames.length > 1 ? '\n' + pathNames.map(s => ` - ${s}`).join('\n') : `'${pathNames[0]}`;
  throw new Error(`File or directory does not exist: ${pathNamesPrint}`);
}
