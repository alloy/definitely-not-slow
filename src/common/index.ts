export * from './config';
export * from './db';
export * from './getParsedPackages';
export * from './types';
export * from './utils';
export * from './support';
