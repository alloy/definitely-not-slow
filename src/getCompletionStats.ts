import { CompletionMeasurement } from './types';
import { max, mean, median } from './utils';

export function getCompletionStats(measurements: CompletionMeasurement[]) {
  const durations = measurements.map(m => m.duration);
  const worst = max(measurements, m => m.duration);
  return {
      mean: mean(durations),
      median: median(durations),
      worst,
  };
}
