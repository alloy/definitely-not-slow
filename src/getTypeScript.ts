import * as fs from 'fs';
import { promisify } from 'util';
import { installAll, typeScriptPath } from 'dtslint/bin/installer';
const exists = promisify(fs.exists);

export async function getTypeScript(
  version: string,
  localTypeScriptPath?: string
): Promise<typeof import('typescript')> {
  const path = typeScriptPath(version, localTypeScriptPath);
  if (!await exists(path)) {
    await installAll();
  }
  if (!await exists(path)) {
    throw new Error(`Version ${version} is not available`);
  }
  return import(path);
}
