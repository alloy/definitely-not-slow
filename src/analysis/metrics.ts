import { mean } from '../measure/utils';
import { PackageBenchmarkSummary, Document, config, getPercentDiff } from '../common';
import { isNumber } from 'util';

export interface FormatOptions {
  precision?: number;
  indent?: number;
}

export const enum SignificanceLevel {
  Warning = 'warning',
  Alert = 'alert',
  Awesome = 'awesome',
}

export interface Metric {
  columnName: string;
  sentenceName: string;
  formatOptions?: FormatOptions;
  getValue: (x: Document<PackageBenchmarkSummary>) => number | undefined;
  getSignificance: (
    percentDiff: number,
    before: Document<PackageBenchmarkSummary>,
    after: Document<PackageBenchmarkSummary>
  ) => SignificanceLevel | undefined;
}

export type MetricName =
  | 'typeCount'
  // | 'memoryUsage'
  | 'assignabilityCacheSize'
  | 'subtypeCacheSize'
  | 'identityCacheSize'
  | 'samplesTaken'
  | 'identifierCount'
  | 'completionsMean'
  | 'completionsMedian'
  | 'completionsStdDev'
  | 'quickInfoMean'
  | 'quickInfoMedian'
  | 'quickInfoStdDev'
  | 'completionsWorstMean'
  | 'quickInfoWorstMean';

function defaultGetSignificance(percentDiff: number): SignificanceLevel | undefined {
  if (percentDiff > config.comparison.percentDiffSevereThreshold) {
    return SignificanceLevel.Alert;
  }
  if (percentDiff > config.comparison.percentDiffWarningThreshold) {
    return SignificanceLevel.Warning;
  }
  if (percentDiff < config.comparison.percentDiffGoldStarThreshold) {
    return SignificanceLevel.Awesome;
  }
}

const getInsignificant = () => undefined;

function getSignificanceProportionalTo(proportionalTo: MetricName) {
  return (percentDiff: number, before: Document<PackageBenchmarkSummary>, after: Document<PackageBenchmarkSummary>) => {
    const proportionalToBeforeValue = metrics[proportionalTo].getValue(before);
    const proportionalToAfterValue = metrics[proportionalTo].getValue(after);
    if (typeof proportionalToBeforeValue === 'number' && typeof proportionalToAfterValue === 'number') {
      const proportionalToPercentDiff = getPercentDiff(proportionalToAfterValue, proportionalToBeforeValue);
      return defaultGetSignificance(percentDiff - proportionalToPercentDiff);
    }
  };
}

export const metrics: { [K in MetricName]: Metric } = {
  typeCount: {
    columnName: 'Type count',
    sentenceName: 'type count',
    formatOptions: { precision: 0 },
    getValue: x => x.body.typeCount,
    getSignificance: getSignificanceProportionalTo('identifierCount'),
  },
  assignabilityCacheSize: {
    columnName: 'Assignability cache size',
    sentenceName: 'assignability cache size',
    formatOptions: { precision: 0 },
    getValue: x => x.body.relationCacheSizes && x.body.relationCacheSizes.assignable,
    getSignificance: getSignificanceProportionalTo('identifierCount'),
  },
  subtypeCacheSize: {
    columnName: 'Subtype cache size',
    sentenceName: 'subtype cache size',
    formatOptions: { precision: 0 },
    getValue: x => x.body.relationCacheSizes && x.body.relationCacheSizes.subtype,
    getSignificance: getSignificanceProportionalTo('identifierCount'),
  },
  identityCacheSize: {
    columnName: 'Identity cache size',
    sentenceName: 'identity cache size',
    formatOptions: { precision: 0 },
    getValue: x => x.body.relationCacheSizes && x.body.relationCacheSizes.identity,
    getSignificance: getSignificanceProportionalTo('identifierCount'),
  },
  samplesTaken: {
    columnName: 'Samples taken',
    sentenceName: 'number of samples taken',
    formatOptions: { precision: 0 },
    getValue: x => Math.max(x.body.completions.trials, x.body.quickInfo.trials),
    getSignificance: getInsignificant,
  },
  identifierCount: {
    columnName: 'Identifiers in tests',
    sentenceName: 'number of identifiers present in test files',
    formatOptions: { precision: 0 },
    getValue: x => x.body.testIdentifierCount,
    getSignificance: getInsignificant,
  },
  completionsMean: {
    columnName: 'Mean duration (ms)',
    sentenceName: 'mean duration for getting completions at a position',
    getValue: x => x.body.completions.mean,
    getSignificance: defaultGetSignificance,
  },
  completionsMedian: {
    columnName: 'Median duration (ms)',
    sentenceName: 'median duration for getting completions at a position',
    getValue: x => x.body.completions.median,
    getSignificance: defaultGetSignificance,
  },
  completionsStdDev: {
    columnName: 'Std. deviation (ms)',
    sentenceName: 'standard deviation of the durations for getting completions at a position',
    getValue: x => x.body.completions.standardDeviation,
    getSignificance: getInsignificant,
  },
  completionsWorstMean: {
    columnName: 'Worst duration (ms)',
    sentenceName: 'worst-case duration for getting completions at a position',
    getValue: x => mean(x.body.completions.worst.completionsDurations),
    getSignificance: defaultGetSignificance,
  },
  quickInfoMean: {
    columnName: 'Mean duration (ms)',
    sentenceName: 'mean duration for getting quick info at a position',
    getValue: x => x.body.quickInfo.mean,
    getSignificance: defaultGetSignificance,
  },
  quickInfoMedian: {
    columnName: 'Median duration (ms)',
    sentenceName: 'median duration for getting quick info at a position',
    getValue: x => x.body.quickInfo.median,
    getSignificance: defaultGetSignificance,
  },
  quickInfoStdDev: {
    columnName: 'Std. deviation (ms)',
    sentenceName: 'standard deviation of the durations for getting quick info at a position',
    getValue: x => x.body.quickInfo.standardDeviation,
    getSignificance: getInsignificant,
  },
  quickInfoWorstMean: {
    columnName: 'Worst duration (ms)',
    sentenceName: 'worst-case duration for getting quick info at a position',
    getValue: x => mean(x.body.quickInfo.worst.quickInfoDurations),
    getSignificance: defaultGetSignificance,
  },
};

export function getInterestingMetrics(before: Document<PackageBenchmarkSummary>, after: Document<PackageBenchmarkSummary>) {
  return Object.values(metrics).reduce((acc: { metric: Metric, percentDiff: number, significance: SignificanceLevel }[], metric) => {
    const aValue = metric.getValue(before);
    const bValue = metric.getValue(after);
    const percentDiff = isNumber(aValue) && isNumber(bValue) && getPercentDiff(bValue, aValue);
    const significance = typeof percentDiff === 'number' && metric.getSignificance(percentDiff, before, after);
    if (percentDiff && significance) {
      return [
        ...acc,
        { metric, percentDiff, significance },
      ];
    }
    return acc;
  }, []);
}
