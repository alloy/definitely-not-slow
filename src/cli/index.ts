import { deserializeArgs } from '../common';
import { benchmark } from './benchmark';
import { getPackagesToBenchmark } from './getPackagesToBenchmark';

if (!module.parent) {
  const entry = process.argv[2];
  const args = deserializeArgs(process.argv.slice(3));
  (async () => {
    try {
      switch (entry) {
        case 'benchmark':
          return benchmark(args);
        case 'getPackagesToBenchmark':
          return getPackagesToBenchmark(args);
        default:
          console.error(`Unrecognized entry '${entry}'`);
      }
    } catch (err) {
      console.error(err);
      process.exit(1);
    }
  })();

  process.on('unhandledRejection', err => {
    console.error(err);
    process.exit(1);
  });
}
