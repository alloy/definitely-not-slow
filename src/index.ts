import * as os from 'os';
import * as path from 'path';
import { measurePerf } from './measurePerf';
import { getTypeScript } from './getTypeScript';
import { getMeasurementStats } from './getMeasurementStats';

if (!module.parent) {
  (async () => {
      const packageName = process.argv[2];
      console.log(`Measuring ${packageName}`);
      const measurement = measurePerf({
          packageName,
          iterations: 1,
          ts: await getTypeScript('next'),
          definitelyTypedRootPath: path.resolve(__dirname, '../../DefinitelyTyped'),
      });
      const completionStats = getMeasurementStats(measurement.completions);
      const quickInfoStats = getMeasurementStats(measurement.quickInfo);
      const { worst: worstCompletion } = completionStats;
      const { worst: worstQuickInfo } = quickInfoStats;
      console.log(os.EOL);
      console.log('  Type count: ', measurement.typeCount);
      console.log('  Cache sizes');
      console.log('    Assignability: ', measurement.relationCacheSizes.assignable);
      console.log('    Identity:      ', measurement.relationCacheSizes.identity);
      console.log('    Subtype:       ', measurement.relationCacheSizes.subtype);
      console.log('  Completions');
      console.log('    Mean (ms):   ', completionStats.mean.toPrecision(6));
      console.log('    Median (ms): ', completionStats.median.toPrecision(6));
      console.log('    Worst');
      console.log('      Duration (ms):  ', worstCompletion.duration.toPrecision(6));
      console.log('      Std. deviation: ', worstCompletion.standardDeviation.toPrecision(6));
      console.log('      Identifier:     ', worstCompletion.identifierText);
      console.log('      Location:       ', `${worstCompletion.fileName}(${worstCompletion.line},${worstCompletion.offset})`);
      console.log('  Quick Info');
      console.log('    Mean (ms):   ', quickInfoStats.mean.toPrecision(6));
      console.log('    Median (ms): ', quickInfoStats.median.toPrecision(6));
      console.log('    Worst');
      console.log('      Duration (ms):  ', worstQuickInfo.duration.toPrecision(6));
      console.log('      Std. deviation: ', worstQuickInfo.standardDeviation.toPrecision(6));
      console.log('      Identifier:     ', worstQuickInfo.identifierText);
      console.log('      Location:       ', `${worstQuickInfo.fileName}(${worstQuickInfo.line},${worstQuickInfo.offset})`);
  })();
}
