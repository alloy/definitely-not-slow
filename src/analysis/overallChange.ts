import { PackageBenchmarkSummary, Document } from '../common';
import { getInterestingMetrics, SignificanceLevel } from './metrics';
import { assertNever } from 'types-publisher/bin/util/util';

type BeforeAndAfter = [Document<PackageBenchmarkSummary> | undefined, Document<PackageBenchmarkSummary>];

export enum OverallChange {
  Same = 0,
  Worse = 1 << 0,
  Better = 1 << 1,
  Mixed = Worse | Better,
}

export function getOverallChangeForSingleComparison(before: Document<PackageBenchmarkSummary>, after: Document<PackageBenchmarkSummary>) {
  let change = OverallChange.Same;
  for (const { significance } of getInterestingMetrics(before, after)) {
    switch (significance) {
      case SignificanceLevel.Alert:
      case SignificanceLevel.Warning:
        change |= OverallChange.Worse;
        break;
      case SignificanceLevel.Awesome:
        change |= OverallChange.Better;
        break;
      default:
        assertNever(significance);
    }
  }
  return change;
}

export function getOverallChangeForComparisons(comparisons: BeforeAndAfter[]): OverallChange {
  let change = OverallChange.Same;
  for (const comparison of comparisons) {
    change |= comparison[0] ? getOverallChangeForSingleComparison(comparison[0], comparison[1]) : OverallChange.Same;
  }
  return change;
}