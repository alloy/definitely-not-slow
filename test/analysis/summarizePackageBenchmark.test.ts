import { summarize } from '../../src/analysis';
const benchmark = require('./fixtures/abbrev.packageBenchmark.json');

describe('analysis', () => {
  describe('summarizePackageBenchmarks', () => {
    test('snapshot', () => {
      expect(summarize(benchmark)).toMatchSnapshot();
    });
  });
});
