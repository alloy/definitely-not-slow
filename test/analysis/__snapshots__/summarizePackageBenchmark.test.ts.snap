// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analysis summarizePackageBenchmarks snapshot 1`] = `
Object {
  "batchRunStart": "2019-06-05T21:02:57.283Z",
  "benchmarkDuration": 6146.653569,
  "completions": Object {
    "mean": 148.38288194,
    "meanCoefficientOfVariation": 0.23178183452633877,
    "median": 146.9316702,
    "standardDeviation": 14.526224099473964,
    "trials": 20,
    "worst": Object {
      "completionsDurations": Array [
        260.302261,
        161.823348,
        209.062858,
        138.921087,
        123.359843,
      ],
      "end": 310,
      "fileName": "types/abbrev/abbrev-tests.ts",
      "identifierText": "abbrev",
      "line": 12,
      "offset": 15,
      "quickInfoDurations": Array [
        144.181488,
        128.053511,
        181.161561,
        116.247208,
        104.89903,
      ],
      "start": 304,
    },
  },
  "languageServiceCrashed": false,
  "memoryUsage": 33111296,
  "packageName": "abbrev",
  "packageVersion": "*",
  "quickInfo": Object {
    "mean": 140.33250924,
    "meanCoefficientOfVariation": 0.21107195517210994,
    "median": 139.3310676,
    "standardDeviation": 12.553366997435981,
    "trials": 20,
    "worst": Object {
      "completionsDurations": Array [
        188.397316,
        111.76699,
        154.360542,
        137.790343,
        142.34316,
      ],
      "end": 275,
      "fileName": "types/abbrev/abbrev-tests.ts",
      "identifierText": "ReadonlyArray",
      "line": 11,
      "offset": 14,
      "quickInfoDurations": Array [
        209.811926,
        229.587525,
        166.220571,
        111.510644,
        173.78985,
      ],
      "start": 262,
    },
  },
  "relationCacheSizes": Object {
    "assignable": 104,
    "identity": 0,
    "subtype": 0,
  },
  "requestedLanguageServiceTestIterations": 5,
  "sourceVersion": "d1039d660b82d765804f731255279f15693b43c6",
  "testIdentifierCount": 20,
  "typeCount": 2044,
  "typeScriptVersion": "3.6.0-dev.20190604",
  "typeScriptVersionMajorMinor": "3.6",
}
`;
